browser/chat/chat_tab_helper.cc

```diff
@@ -0,0 +1,207 @@
+
+#include "brave/browser/chat/chat_tab_helper.h"
+
+#include <queue>
+#include <utility>
+
+#include "base/strings/strcat.h"
+#include "brave/components/chat_ui/common/chat_ui_constants.h"
+#include "ui/accessibility/ax_node.h"
+#include "ui/accessibility/ax_tree.h"
+
+namespace {
+static const ax::mojom::Role kContentRoles[]{
+    ax::mojom::Role::kHeading,
+    ax::mojom::Role::kParagraph,
+};
+
+static const ax::mojom::Role kRolesToSkip[]{
+    ax::mojom::Role::kAudio,
+    ax::mojom::Role::kBanner,
+    ax::mojom::Role::kButton,
+    ax::mojom::Role::kComplementary,
+    ax::mojom::Role::kContentInfo,
+    ax::mojom::Role::kFooter,
+    ax::mojom::Role::kFooterAsNonLandmark,
+    ax::mojom::Role::kImage,
+    ax::mojom::Role::kLabelText,
+    ax::mojom::Role::kNavigation,
+};
+
+void GetContentRootNodes(const ui::AXNode* root,
+                         std::vector<const ui::AXNode*>* content_root_nodes) {
+  std::queue<const ui::AXNode*> queue;
+  queue.push(root);
+  while (!queue.empty()) {
+    const ui::AXNode* node = queue.front();
+    queue.pop();
+    // If a main or article node is found, add it to the list of content root
+    // nodes and continue. Do not explore children for nested article nodes.
+    if (node->GetRole() == ax::mojom::Role::kMain ||
+        node->GetRole() == ax::mojom::Role::kArticle) {
+      content_root_nodes->push_back(node);
+      continue;
+    }
+    for (auto iter = node->UnignoredChildrenBegin();
+         iter != node->UnignoredChildrenEnd(); ++iter) {
+      queue.push(iter.get());
+    }
+  }
+}
+
+void AddContentNodesToVector(const ui::AXNode* node,
+                             std::vector<const ui::AXNode*>* content_nodes) {
+  if (base::Contains(kContentRoles, node->GetRole())) {
+    content_nodes->emplace_back(node);
+    return;
+  }
+  if (base::Contains(kRolesToSkip, node->GetRole())) {
+    return;
+  }
+  for (auto iter = node->UnignoredChildrenBegin();
+       iter != node->UnignoredChildrenEnd(); ++iter) {
+    AddContentNodesToVector(iter.get(), content_nodes);
+  }
+}
+
+void AddTextNodesToVector(const ui::AXNode* node,
+                          std::vector<std::u16string>* strings) {
+  const ui::AXNodeData& node_data = node->data();
+
+  if (node_data.role == ax::mojom::Role::kStaticText) {
+    if (node_data.HasStringAttribute(ax::mojom::StringAttribute::kName)) {
+      strings->emplace_back(
+          node_data.GetString16Attribute(ax::mojom::StringAttribute::kName));
+    }
+    return;
+  }
+
+  for (const auto* child : node->children()) {
+    AddTextNodesToVector(child, strings);
+  }
+}
+}  // namespace
+
+ChatTabHelper::ChatTabHelper(content::WebContents* web_contents)
+    : content::WebContentsObserver(web_contents),
+      content::WebContentsUserData<ChatTabHelper>(*web_contents) {}
+
+ChatTabHelper::~ChatTabHelper() = default;
+
+std::string ChatTabHelper::GetConversationHistoryAsString() {
+  std::string history_text;
```

Comment:
I think you might want to cache this |history_text| too. Then you could also change return type to `const std::string&`

Reply:
Do we want to cache the history text because using a const reference requires an initial value?

Reply:
I think I prefer avoiding string manipulation when we don't need.

Reply:
Done